### ISSUE TITLE
Soft delete character for easy restoration

### ISSUE THREAD

**Issue by nasomi** [DarkstarProject/darkstar#4607]
https://github.com/DarkstarProject/darkstar/blob/master/src/login/lobby.cpp#L547

Replace with:
`const char *pfmtQuery = "update chars set charname = concat(charname, 'deleted'), accid = concat(999,accid) where charid = %i";
`

**Comment by TeoTwawki** [comment]
Suggestion:
`concat(charname, '_deleted')` so Timmy becomes Timmy_deleted instead of Timmydeleted

**Comment by teschnei** [comment]
probably would be best to just have a new column in chars for deleted or not

**Comment by TeoTwawki** [comment]
Depends on if you want to free up the name or not I guess

**Comment by nasomi** [comment]
Freeing up the charname is the whole point of deleting character. People will make and remake chars lots of times using the same name but changing appearance.

**Comment by kamajii7** [comment]
Not sure if this is my place, but to make it unique shouldn't we add the character id? i.e. Timmy_10_deleted. I just see the issue of freeing up a name then someone else taking the name then another soft delete happening, causing two Timmy_deleted's, may cause confusion for administrators.

**Comment by teschnei** [comment]
Better solution is to add a deleted column, and upon deletion, move the
name from name to deleted. Keeps the name for querying later, but takes it
out of the actual name column for uniqueness.

**Comment by kamajii7** [comment]
I completely agree. I am just thinking of the interim for a quick solution.

**Comment by nasomi** [comment]
Yeah this does not need to be this complicated. I’ve been running this system for 2 years now and it’s served every undelelete I needed. A distinct name is hardly required for deleted chars, you could just look up the unmodofied account id to see the login. The person in question should be able to provide the login name for the char they want Undeleted.

**Comment by takhlaq** [comment]
an additional field in chars will do
is_deleted

**Comment by neuromancerxi** [comment]
teschnei and takhlaq suggestion/solution is nice since it doesn't require modification to the name field, and accomplishes the task with a bit change instead of a string change. One wrench in the works by using the extra field method is that some of the tertiary things that interface with the db might be impacted by this change (Arcanus, Afterhours AH, DSP Web, etc) when doing a lookup by character name. 

Deleted character records would still be uniquely identified via charid/accid, but until tertiary applications update their logic to look for an additional flag to delete/exclude/do-something-else, there may be some confusion in query results.

You'd need a manual process/procedure to deal with a name collision if the character name has been re-used between deletion and restoration request (think about malicious or account takeover scenarios), but that's out of scope for this conversation.

As an aside, I'm not aware of an un-delete method in Retail that allows a user to restore a character without [contacting customer support](http://support.na.square-enix.com/faqarticle.php?kid=12457&id=20&la=1&ret=faq&pv=20&page=1&c=0&sc=0&so=0), so this would still be in line with how the Retail world does things.

**Comment by takhlaq** [comment]
they shouldnt be affected if the field is added on to the end, even if external tools 
```sql
SELECT * FROM chars
```

instead of selecting each field by name

**Comment by neuromancerxi** [comment]
It depends on how their query is being built:

```sql
SELECT charid,accid,charname FROM chars WHERE charname like "something%"```

This would return multiple hits for different characters instead of a unique result by name as it does today. This isn't world ending, the query could be modified to include an additional qualifier:

```sql
SELECT charid,accid,charname FROM chars WHERE charname like "something%" AND is_deleted != 1```

Just wanted to bring it up as something to consider as a repercussion of this proposed change.

**Comment by xipies** [comment]
The two routes I have gone on other projects are:

* A flag, like is_deleted, and all queries had to be updated and/or rerouted (e.g., through a filtered view).
* Renamed the entity to a random value (uuid), and if needed (was not always), stored the original name somewhere else, such as another column. This was done where I less control over the usage.
